File: 5.longest-palindromic-substring.py
#
# @lc app=leetcode id=5 lang=python3
#
# [5] Longest Palindromic Substring
#


# @lc code=start
class Solution:
    def longestPalindrome(self, s: str) -> str:
        n = len(s)

        ret = ''
        for i in range(n):
            # 奇回文的情况
            l, r = i, i
            while s[l] == s[r]:
                l -= 1
                r += 1
                if l < 0 or r >= n:
                    break
            if r - l - 1 > len(ret):
                ret = s[l + 1:r]

            # 偶回文的情况
            l, r = i - 1, i
            while l >= 0 and s[l] == s[r]:
                l -= 1
                r += 1
                if l < 0 or r >= n:
                    break
            if r - l - 1 > len(ret):
                ret = s[l + 1:r]

        return ret


# @lc code=end

File: 5.longest-palindromic-substring.go
package leetcode

// longestPalindrome TODO
// 5. 最长回文子串
func longestPalindrome(s string) string {
	n := len(s)
	if n <= 1 {
		return s
	}
	D := make([][]bool, n)
	for i := range D {
		D[i] = make([]bool, n)
	}
	// init: 长度为 1 的子串都是回文
	for i := 0; i < n; i++ {
		D[i][i] = true
	}
	maxLength := 1
	start := 0
	// 状态转移方程，D[i][j] = (D[i+1][j-1] || 少于等于两个字符) && s[i] == s[j]
	for L := 2; L <= n; L++ { // 从长度较短的字符串向长度较长的字符串进行转移的，因此一定要注意动态规划的循环顺序。
		for i := 0; i < n; i++ {
			j := i + L - 1 // j - i + 1 = L
			if j >= n {
				break
			}

			if s[i] == s[j] && (j-i <= 2 || D[i+1][j-1]) {
				D[i][j] = true
				if maxLength < L {
					maxLength = L
					start = i
				}
			}
		}
	}
	return s[start : start+maxLength]
}
